daemon/health.go

```diff
@@ -248,12 +248,37 @@ func handleProbeResult(d *Daemon, c *container.Container, result *types.Healthch
 // There is never more than one monitor thread running per container at a time.
 func monitor(d *Daemon, c *container.Container, stop chan struct{}, probe probe) {
 	probeInterval := timeoutWithDefault(c.Config.Healthcheck.Interval, defaultProbeInterval)
+	startInterval := timeoutWithDefault(c.Config.Healthcheck.StartInterval, defaultProbeInterval)
+	startPeriod := timeoutWithDefault(c.Config.Healthcheck.StartPeriod, defaultStartPeriod)
 
-	intervalTimer := time.NewTimer(probeInterval)
+	c.Lock()
+	started := c.State.StartedAt
+	c.Unlock()
+
+	getInterval := func() time.Duration {
+		if time.Now().Sub(started) >= startPeriod {
+			return probeInterval
+		}
+		c.Lock()
+		status := c.Health.Health.Status
+		c.Unlock()
+
+		if status == types.Starting {
+			return startInterval
+		}
+		return probeInterval
+	}
+
+	intervalTimer := time.NewTimer(0)
 	defer intervalTimer.Stop()
 
+	// We aren't using the timer yet so stop it and drain the channel so it is ready when we hit the loop.
+	if !intervalTimer.Stop() {
+		<-intervalTimer.C
+	}
+
 	for {
-		intervalTimer.Reset(probeInterval)
+		intervalTimer.Reset(getInterval())
```

Comment:
Sure, that works, though the stop-and-drain dance wouldn't be needed if the reset was moved to the bottom of the loop.
```go
intervalTimer := time.NewTimer(getInterval())
defer intervalTimer.Stop()

for {
	select { /* ... */ }
	intervalTimer.Reset(probeInterval())
}
```

Reply:
Fixed that as well.